diff --git a/CMakeLists.txt b/CMakeLists.txt
index 167ce19578..da220cc5f8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -284,6 +284,9 @@ OCV_OPTION(WITH_WAYLAND "Include Wayland support" OFF
 OCV_OPTION(WITH_IPP "Include Intel IPP support" (NOT MINGW AND NOT CV_DISABLE_OPTIMIZATION)
   VISIBLE_IF (X86_64 OR X86) AND NOT WINRT AND NOT IOS AND NOT XROS
   VERIFY HAVE_IPP)
+OCV_OPTION(WITH_ARMPL "Include ARM Performance Libraries support" (NOT CV_DISABLE_OPTIMIZATION)
+  VISIBLE_IF (AARCH64 OR ARM64) AND NOT WINRT AND NOT IOS AND NOT XROS
+  VERIFY HAVE_ARMPL)
 OCV_OPTION(WITH_HALIDE "Include Halide support" OFF
   VISIBLE_IF TRUE
   VERIFY HAVE_HALIDE)
@@ -1783,6 +1786,12 @@ if(WITH_IPP AND HAVE_IPP)
   endif()
 endif()
 
+if(WITH_ARMPL AND HAVE_ARMPL)
+  status("    ARM Perf Lib:" "${ARMPL_VERSION_STR}")
+  status("           at:" "${ARMPL_ROOT_DIR}")
+  status("      variant:" "${ARMPL_LIB_NAME}")
+endif()
+
 if(WITH_VA OR HAVE_VA)
   status("    VA:"            HAVE_VA          THEN "YES" ELSE NO)
 endif()
diff --git a/cmake/OpenCVFindLibsPerf.cmake b/cmake/OpenCVFindLibsPerf.cmake
index 861a39c47f..19bf5ac9cd 100644
--- a/cmake/OpenCVFindLibsPerf.cmake
+++ b/cmake/OpenCVFindLibsPerf.cmake
@@ -38,6 +38,20 @@ if(WITH_IPP)
   endif()
 endif()
 
+if(WITH_ARMPL)
+  include("${OpenCV_SOURCE_DIR}/cmake/OpenCVFindArmPL.cmake")
+  if(HAVE_ARMPL)
+    message(STATUS "Using ARM Performance Libraries")
+    ocv_include_directories(${ARMPL_INCLUDE_DIRS})
+    list(APPEND OPENCV_LINKER_LIBS ${ARMPL_LIBRARIES})
+    add_compile_definitions(HAVE_ARMPL)
+    
+    if(WIN32 AND NOT BUILD_SHARED_LIBS)
+      add_compile_definitions(ARMPL_STATIC)
+    endif()
+  endif()
+endif()
+
 # --- CUDA ---
 if(WITH_CUDA)
   if(ENABLE_CUDA_FIRST_CLASS_LANGUAGE)
